sdk/storage/azure-storage-blob/src/main/java/com/azure/android/storage/blob/transfer/ReadableContent.java

```diff
@@ -0,0 +1,203 @@
+        this.context = context;
+        this.contentUri = contentUri;
+        this.useContentResolver = useContentResolver;
+    }
+
+    /**
+     * Get the {@link Uri} to the content.
+     *
+     * @return the content URI
+     */
+    Uri getUri() {
+        return this.contentUri;
+    }
+
+    /**
+     * Check whether to use {@link android.content.ContentResolver} to resolve the content URI.
+     *
+     * @return true if resolving content URI requires content resolver.
+     */
+    boolean isUseContentResolver() {
+        return this.useContentResolver;
+    }
+
+    /**
+     * Attempt to take persistable read permission on the content.
+     *
+     * @throws IllegalStateException if read permission is not granted
+     */
+    @MainThread
+    void takePersistableReadPermission() throws IllegalStateException {
+        if (this.useContentResolver) {
+            this.context.getContentResolver()
+                .takePersistableUriPermission(this.contentUri, Intent.FLAG_GRANT_READ_URI_PERMISSION);
+            checkPersistableReadGranted();
+        }
+    }
+
+    /**
+     * Get the total size of the content in bytes.
+     *
+     * @return the content size in bytes
+     * @throws Throwable if retrieval of content length fails
+     */
+    long getLength() throws Throwable {
+        final long contentLength;
+        if (this.useContentResolver) {
+            // https://developer.android.com/training/secure-file-sharing/retrieve-info
+            final Cursor cursor = this.context.getContentResolver().query(this.contentUri, null, null, null, null);
+            final int sizeIndex = cursor.getColumnIndex(OpenableColumns.SIZE);
+            cursor.moveToFirst();
+            contentLength = cursor.getLong(sizeIndex);
+        } else {
+            File file = new File(contentUri.getPath());
+            contentLength = file.length();
+        }
+        if (contentLength == -1) {
+            throw new Throwable("Unable to get size of the content '" + contentUri + "'.");
+        }
+        return contentLength;
+    }
+
+    /**
+     * Read a block of bytes from the content.
+     *
+     * @param blockOffset the start offset of the block
+     * @param blockSize the size of the block
+     * @return block of bytes in the range [blockOffset, blockOffset + blockSize]
+     * @throws IOException the IO error when attempting to read
+     * @throws IllegalStateException if read permission is not granted or revoked
+     */
+    byte[] readBlock(int blockOffset, int blockSize) throws IOException, IllegalStateException {
+        if (this.useContentResolver) {
+            this.checkPersistableReadGranted();
+            // ContentResolver::openFileDescriptor works but we use openAssetFileDescriptor
+            // so that providers that returns subsection of file gets supported.
+            // The "r" (read) mode is used so that the content providers that don't support write can also consumed.
```

Comment:
```suggestion
            // so that providers that return subsections of a file are supported.
            // The "r" (read) mode is used so that the content providers that don't support write can also be consumed.
```